ext/uri/php_uri.c

```diff
@@ -106,6 +106,224 @@ static HashTable *uri_get_debug_properties(zend_object *object)
+
+	if (UNEXPECTED(internal_uri->uri == NULL)) {
+		efree(internal_uri);
+		return NULL;
+	}
+
+	return internal_uri;
+}
+
+static zend_result php_uri_get_property(const uri_internal_t *internal_uri, uri_property_name_t property_name, uri_component_read_mode_t read_mode, zval *zv)
+{
+	const uri_property_handler_t *property_handler = uri_property_handler_from_internal_uri(internal_uri, property_name);
+	if (property_handler == NULL) {
+		return FAILURE;
+	}
+
+	return property_handler->read_func(internal_uri, read_mode, zv);
+}
+
+PHPAPI zend_result php_uri_get_scheme(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *zv)
+{
+	return php_uri_get_property(internal_uri, URI_PROPERTY_NAME_SCHEME, read_mode, zv);
+}
+
+PHPAPI zend_result php_uri_get_username(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *zv)
+{
+	return php_uri_get_property(internal_uri, URI_PROPERTY_NAME_USERNAME, read_mode, zv);
+}
+
+PHPAPI zend_result php_uri_get_password(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *zv)
+{
+	return php_uri_get_property(internal_uri, URI_PROPERTY_NAME_PASSWORD, read_mode, zv);
+}
+
+PHPAPI zend_result php_uri_get_host(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *zv)
+{
+	return php_uri_get_property(internal_uri, URI_PROPERTY_NAME_HOST, read_mode, zv);
+}
+
+PHPAPI zend_result php_uri_get_port(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *zv)
+{
+	return php_uri_get_property(internal_uri, URI_PROPERTY_NAME_PORT, read_mode, zv);
+}
+
+PHPAPI zend_result php_uri_get_path(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *zv)
+{
+	return php_uri_get_property(internal_uri, URI_PROPERTY_NAME_PATH, read_mode, zv);
+}
+
+PHPAPI zend_result php_uri_get_query(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *zv)
+{
+	return php_uri_get_property(internal_uri, URI_PROPERTY_NAME_QUERY, read_mode, zv);
+}
+
+PHPAPI zend_result php_uri_get_fragment(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *zv)
+{
+	return php_uri_get_property(internal_uri, URI_PROPERTY_NAME_FRAGMENT, read_mode, zv);
+}
+
+PHPAPI void php_uri_free(uri_internal_t *internal_uri)
+{
+	internal_uri->handler->free_uri(internal_uri->uri);
+	internal_uri->uri = NULL;
+	internal_uri->handler = NULL;
+	efree(internal_uri);
+}
+
+PHPAPI php_uri *php_uri_parse_to_struct(const uri_handler_t *uri_handler, zend_string *uri_str, uri_component_read_mode_t read_mode, bool silent)
+{
+	uri_internal_t *uri_internal = php_uri_parse(uri_handler, uri_str, silent);
+	if (uri_internal == NULL) {
+		return NULL;
+	}
+
+	php_uri *uri = ecalloc(1, sizeof(*uri));
+	zval tmp;
+	ZVAL_UNDEF(&tmp);
+	zend_result result;
+
+	result = php_uri_get_scheme(uri_internal, read_mode, &tmp);
+	if (result == FAILURE) {
+		php_uri_free(uri_internal);
```

Comment:
Is it possible to avoid the repeated error blocks by using a single "error" label that you goto to?
I'd say one of the only proper use-cases for goto in C is for these kinds of repeated error handlings.